From: Matthias Clasen <mclasen@redhat.com>
Date: Thu, 11 Sep 2025 14:57:07 -0400
Subject: Fix upright text carets

This fixes an issue introduced in 2c7073b86b3c4a2d5.

By using inclusive rounding for the caret position rectangle,
wevinadvertendly turn upright carets (with a width of zero)
into slanted carets. Don't do that.

Issue found by Adam Williamson.

Part-of: <https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/8933>
(cherry picked from commit 7060039ec59d31a8b5279f648e5bf91731544627)

Origin: upstream, after 4.20.1
---
 gtk/gtkrenderlayout.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/gtk/gtkrenderlayout.c b/gtk/gtkrenderlayout.c
index 102663f..263ff6d 100644
--- a/gtk/gtkrenderlayout.c
+++ b/gtk/gtkrenderlayout.c
@@ -290,6 +290,7 @@ gtk_css_style_snapshot_caret (GtkCssBoxes    *boxes,
   GdkSeat *seat;
   PangoDirection keyboard_direction;
   PangoDirection direction2;
+  gboolean width_was_zero;
 
   g_object_get (gtk_settings_get_for_display (display),
                 "gtk-split-cursor", &split_cursor,
@@ -307,8 +308,19 @@ gtk_css_style_snapshot_caret (GtkCssBoxes    *boxes,
     }
 
   pango_layout_get_caret_pos (layout, index, &strong_pos, &weak_pos);
+
+  /* We special-case a width of zero here, because we don't want
+   * an upright cursor go sloped due to rounding.
+   */
+  width_was_zero = strong_pos.width == 0;
   pango_extents_to_pixels (&strong_pos, NULL);
+  if (width_was_zero)
+    strong_pos.width = 0;
+
+  width_was_zero = weak_pos.width == 0;
   pango_extents_to_pixels (&weak_pos, NULL);
+  if (width_was_zero)
+    weak_pos.width = 0;
 
   direction2 = PANGO_DIRECTION_NEUTRAL;
   cursor2 = NULL; /* poor MSVC */
